src/Access/IAccessStorage.cpp

```diff
@@ -39,6 +43,29 @@ namespace
     {
         return "ID(" + toString(id) + ")";
     }
+
+    std::shared_ptr<const EnabledQuota> getAuthenticationQuota(UUID user_id,
+                                                               const UserPtr & user,
+                                                               const Poco::Net::IPAddress & address,
+                                                               const std::string & forwarded_address)
+    {
+        /// During authentication process, client_key is not received from TCP client,
+        /// Use predefined authentication quota key and always receive the save Interval object
+        /// to avoid throwing exceptions in case of QuotaKeyType::CLIENT_KEY key type.
+        constexpr auto AUTHENTICATION_QUOTA_KEY = "_AUTHENTICATION_QUOTA_KEY_";
+        const auto new_current_roles = user->granted_roles.findGranted(user->default_roles);
+        const auto & access = Context::getGlobalContextInstance()->getAccessControl();
```

Comment:
Scary in this line.

Reply:
@alexey-milovidov I discussed with @Demilivor the line No 57 is really scary. It should to be refactored. 
It seems like many parts of code related to quotas were changed during the past 3 months. We probably need more guidance about core ClickHouse, Inc developers about implementation of the "maximum sequential login failures" feature.

Right now @Demilivor was on another project but he will return back on this PR after we discuss it with our management. The task is actual for our company.

Reply:
I moved this code to another place, and now my code does not call `Context::getGlobalContextInstance()->getAccessControl()`